### PR TITLE
ROX-19828: Add display names to vulnerability requests

### PR DESCRIPTION
**Pull request by md2119**
## Description

This PR adds system-generated display names to vulnerability requests and migration to set names for existing vulnerability requests. The display name follows `{user-short-name}-{yymmdd}-{seq-num}` format. The `seq-num` sequence is reset when the month changes to keep the sequence number low. When generating the names for existing requests in migration, the sequence numbers are not necessarily in the order of the creation time to avoid search and ordering overhead.

The following illustrates how the display name would look like for vuln requests created in order:
`AD-220601-1` - first request on 06/01/2023
`AD-220601-2` - second request on on 06/01/2023
`MD-230801-1` - first request  on 08/01/2023
`MD-230811-2` - second request  on 08/11/2023
`MD-230821-3` - third request  on 08/21/2023
`ZD-230901-1` - first request  on 09/01/2023

Recommendation for users to review commits separately. The first commit updates the workflow and the second commit adds migration.

## Checklist
- [ ] Investigated and inspected CI test results
- [X] Unit test and regression tests added
- ~Evaluated and added CHANGELOG entry if required~
- ~Determined and documented upgrade steps~
- ~Documented user facing changes ~
If any of these don't apply, please comment below.

### Here I tell how I validated my change
Unit tests

### CODE REVIEW

File: central/graphql/resolvers/vulnerability_requests.go [hunk 1]
```diff
@@ -89,6 +90,7 @@ func (resolver *Resolver) DeferVulnerability(
 			return nil, err
 		}
 		req := vulnReqUtils.V1DeferVulnRequestToVulnReq(ctx, args.Request.AsV1DeferralRequest())
+		fmt.Println("1")
```
Comment (theencee) [inline central/graphql/resolvers/vulnerability_requests.go:93]: remove

File: central/vulnerabilityrequest/manager/requestmgr/manager_impl.go [hunk 2]
```diff
@@ -105,7 +120,8 @@ func (m *managerImpl) Create(ctx context.Context, req *storage.VulnerabilityRequ
 			fmt.Sprintf("%s already covered by request %s for the scope", errPrefix, existingReqs[idx].GetId()))
 	}
 
-	req.Id = uuid.NewV4().String()
+	m.setIDAndName(req)
+	fmt.Printf("name=%s\n", req.GetName())
```
Comment (theencee) [inline central/vulnerabilityrequest/manager/requestmgr/manager_impl.go:124]: remove

File: central/vulnerabilityrequest/manager/requestmgr/manager_impl.go [hunk 3]
```diff
@@ -572,3 +630,54 @@ func getReprocessDeploymentMsg(deps ...string) *central.MsgToSensor {
...
+	name := strings.ToUpper(user.GetName())
+	parts := strings.Split(name, " ")
+	for i := 0; i < len(parts); i++ {
+		parts[i] = strings.TrimSpace(parts[i])
+	}
+
+	firstName := stringutils.FirstNonEmpty(parts...)
+	lastName := stringutils.LastNonEmpty(parts...)
+	if firstName != "" && lastName != "" {
+		return fmt.Sprintf("%c%c", firstName[0], lastName[0])
```
Comment (theencee) [inline central/vulnerabilityrequest/manager/requestmgr/manager_impl.go:680]: Names are complicated things and there really isn't a good way of getting initials. For someone with only a first name it'll always show SYS (the default). And some folks consider their last name, their first, etc. 

Since it doesn't actually matter that the short name is an initial, why not take the first 2 or 3 chars from the name itself (with spaces removed)?

File: pkg/search/postgres/store.go [hunk 4]
```diff
@@ -95,14 +97,24 @@ func NewGenericStoreWithPermissionChecker[T any, PT unmarshaler[T]](
...
-		insertInto:                       insertInto,
-		setAcquireDBConnDuration:         setAcquireDBConnDuration,
-		setPostgresOperationDurationTime: setPostgresOperationDurationTime,
-		permissionChecker:                checker,
+		db:          db,
+		schema:      schema,
+		pkGetter:    pkGetter,
+		copyFromObj: copyFromObj,
+		insertInto:  insertInto,
+		setAcquireDBConnDuration: func() durationTimeSetter {
```
Comment (theencee) [inline pkg/search/postgres/store.go:105]: What's the reason behind this change? When can `setAcquireDBConnDuration` be null?

File: central/vulnerabilityrequest/manager/requestmgr/manager_impl.go (hunk 3 above)
Comment (md2119) [reply central/vulnerabilityrequest/manager/requestmgr/manager_impl.go:680]: For the folks that have one worded name, the first and the last name would be non-empty; note that we are looking for `LastNonEmpty`. `SYS` will be used only if there is some invalid situation like `" "` API token name.

**Review by theencee — APPROVED** [review]

File: pkg/search/postgres/store.go (hunk 4 above)
Comment (md2119) [reply pkg/search/postgres/store.go:105]: We do not log metrics in migrations, but we copy store code that has metric logging.

File: migrator/migrations/m_189_to_m_190_vulnerability_requests_add_name/migration_impl.go [hunk 5]
```diff
@@ -0,0 +1,198 @@
+package m189tom190
```
Comment (dashrews78) [inline migrator/migrations/m_189_to_m_190_vulnerability_requests_add_name/migration_impl.go:1]: A warning here, this is the 3rd in flight PR with this migration number.  Whomever goes first is in good shape, the others will have to renumber.

Comment (md2119) [reply migrator/migrations/m_189_to_m_190_vulnerability_requests_add_name/migration_impl.go:1]: Thanks for the callout

File: migrator/migrations/m_189_to_m_190_vulnerability_requests_add_name/migration_impl.go [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+	return setRequestNames(ctx, db, lastKnownSeqNumInfo)
+}
+
+func collectExistingSeqNums(ctx context.Context, database *gorm.DB) (map[time.Month]int, error) {
+	db := database.WithContext(ctx).Table(newSchema.VulnerabilityRequestsTableName)
+	rows, err := db.Rows()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to iterate table %s", newSchema.VulnerabilityRequestsTableName)
+	}
+	if rows.Err() != nil {
```
Comment (dashrews78) [inline migrator/migrations/m_189_to_m_190_vulnerability_requests_add_name/migration_impl.go:54]: This was pointed out on my PLOP prune PR, this check should go after the loop.  It has to be the last thing done for rows as rows.Err() would be populated if there were issues processing one of the batches.

File: migrator/migrations/m_189_to_m_190_vulnerability_requests_add_name/migration_impl.go [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+	return lastKnownSeqNumInfo, nil
+}
+
+func setRequestNames(ctx context.Context, database *gorm.DB, lastKnownSeqNumInfo map[time.Month]int) error {
+	db := database.WithContext(ctx).Table(newSchema.VulnerabilityRequestsTableName)
+	rows, err := db.Rows()
+	if err != nil {
+		return errors.Wrapf(err, "failed to iterate table %s", newSchema.VulnerabilityRequestsTableName)
+	}
+	if rows.Err() != nil {
```
Comment (dashrews78) [inline migrator/migrations/m_189_to_m_190_vulnerability_requests_add_name/migration_impl.go:91]: same as comment above

File: migrator/migrations/m_189_to_m_190_vulnerability_requests_add_name/migration_test.go [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+	require.NoError(s.T(), err)
+	aug29, err := time.Parse(layout, "2023/08/29")
+	require.NoError(s.T(), err)
+	aug30, err := time.Parse(layout, "2023/08/30")
+	require.NoError(s.T(), err)
+	sept29, err := time.Parse(layout, "2023/09/29")
+	require.NoError(s.T(), err)
+
+	vulnReqs := []*storage.VulnerabilityRequest{
+		testRequestNameFakeVulnReq(aug29, "bruce wayne"),
```
Comment (dashrews78) [inline migrator/migrations/m_189_to_m_190_vulnerability_requests_add_name/migration_test.go:62]: 🦇

File: migrator/migrations/m_189_to_m_190_vulnerability_requests_add_name/store/previous/store.go [hunk 9]
```diff
@@ -0,0 +1,302 @@
+package postgres
```
Comment (dashrews78) [inline migrator/migrations/m_189_to_m_190_vulnerability_requests_add_name/store/previous/store.go:1]: It looks like these are only used for the tests.  I think we should put them in a test package.  either `test/store/previous|updated` or `store/test/previous|updated`

File: migrator/migrations/m_189_to_m_190_vulnerability_requests_add_name/store/previous/store.go [hunk 10]
```diff
@@ -0,0 +1,302 @@
+package previous
```
Comment (dashrews78) [inline migrator/migrations/m_189_to_m_190_vulnerability_requests_add_name/store/previous/store.go:1]: I would put these in a `test` folder to make it clear they are only needed for test and not the migration itself.

**Review by dashrews78 — APPROVED** [review]
Looks good.  Only a comment about putting the stores in a `test` folder to make it clear they are not used in the migration.

File: migrator/migrations/m_189_to_m_190_vulnerability_requests_add_name/store/previous/store.go (hunk 10 above)
Comment (md2119) [reply migrator/migrations/m_189_to_m_190_vulnerability_requests_add_name/store/previous/store.go:1]: I think it is okay, in that, we do not disallow reading through stores.  It is better to stick to one pattern.
